src/sinks/splunk_hec/common/acknowledgements.rs

```diff
@@ -0,0 +1,314 @@
+    fn default() -> Self {
+        Self {
+            query_interval: NonZeroU8::new(10).unwrap(),
+            retry_limit: NonZeroU8::new(30).unwrap(),
+        }
+    }
+}
+
+pub fn default_hec_client_acknowledgements_config() -> Option<HecClientAcknowledgementsConfig> {
+    Some(Default::default())
+}
+
+#[derive(Deserialize, Serialize, Eq, PartialEq, Debug)]
+pub struct HecAckStatusRequest {
+    pub acks: Vec<u64>,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct HecAckStatusResponse {
+    pub acks: HashMap<u64, bool>,
+}
+
+enum HecAckApiError {
+    ClientBuildRequest,
+    ClientParseResponse,
+    ClientSendQuery,
+    ServerSendQuery,
+}
+
+struct HecAckClient {
+    acks: HashMap<u64, (u8, Sender<EventStatus>)>,
+    retry_limit: u8,
+    client: HttpClient,
+    http_request_builder: Arc<HttpRequestBuilder>,
+}
+
+impl HecAckClient {
+    fn new(
+        retry_limit: u8,
+        client: HttpClient,
+        http_request_builder: Arc<HttpRequestBuilder>,
+    ) -> Self {
+        Self {
+            acks: HashMap::new(),
+            retry_limit,
+            client,
+            http_request_builder,
+        }
+    }
+
+    /// Adds an ack id to be queried
+    fn add(&mut self, ack_id: u64, ack_event_status_sender: Sender<EventStatus>) {
+        self.acks
+            .insert(ack_id, (self.retry_limit, ack_event_status_sender));
+    }
+
+    /// Queries Splunk HEC with stored ack ids and finalizes events that are successfully acked
+    async fn run(&mut self) {
+        let ack_query_body = self.get_ack_query_body();
+        if !ack_query_body.acks.is_empty() {
+            let ack_query_response = self.send_ack_query_request(&ack_query_body).await;
+
+            match ack_query_response {
+                Ok(ack_query_response) => {
+                    debug!("Received ack statuses {:?}", ack_query_response);
+                    let acked_ack_ids = ack_query_response
+                        .acks
+                        .iter()
+                        .filter_map(|(ack_id, ack_status)| ack_status.then(|| *ack_id))
+                        .collect::<Vec<u64>>();
+                    self.finalize_delivered_ack_ids(acked_ack_ids.as_slice());
+                    self.expire_ack_ids_with_status(EventStatus::Rejected);
+                }
+                Err(error) => {
+                    match error {
+                        HecAckApiError::ClientParseResponse | HecAckApiError::ClientSendQuery => {
+                            // If we are permanently unable to interact with
+                            // Splunk HEC indexer acknowledgements (e.g. due to
+                            // request/response format changes in future
+                            // versions), log an error and fall back to default
+                            // behavior.
+                            error!(message = "Unable to parse ack query response. Acknowledging based on initial 200 OK.");
```

Comment:
Should this be an internal event with a metric?

Reply:
Sure thing, I'll refactor this and others mentioned to be more consistent with standard internal instrumentation.